llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp

```diff
@@ -0,0 +1,574 @@
+         "Unexpected unconditional jump instruction");
+  assert(StackHeight >= 0);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoJUMP))
+                   .addMBB(Target);
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitCondJump(
+    const MachineInstr *MI, MachineBasicBlock *Target) {
+  assert(MI->getOpcode() == EVM::JUMPI &&
+         "Unexpected conditional jump instruction");
+  StackHeight -= 1;
+  assert(StackHeight >= 0);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoJUMPI))
+                   .addMBB(Target);
+  verify(NewMI);
+}
+
+// Verify that a stackified instruction doesn't have registers and dump it.
+void EVMStackifyCodeEmitter::CodeEmitter::verify(const MachineInstr *MI) const {
+  assert(EVMInstrInfo::isStack(MI) &&
+         "Only stackified instructions are allowed");
+  assert(all_of(MI->operands(),
+                [](const MachineOperand &MO) { return !MO.isReg(); }) &&
+         "Registers are not allowed in stackified instructions");
+
+  LLVM_DEBUG(dbgs() << "Adding: " << *MI << "stack height: " << StackHeight
+                    << "\n");
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::finalize() {
+  for (MachineBasicBlock &MBB : MF)
+    for (MachineInstr &MI : make_early_inc_range(MBB))
+      // Remove all the instructions that are not stackified.
+      // FIXME: Fix debug info for stackified instructions and don't
+      // remove debug instructions.
+      if (!EVMInstrInfo::isStack(&MI))
+        MI.eraseFromParent();
+
+  auto *MFI = MF.getInfo<EVMMachineFunctionInfo>();
+  MFI->setIsStackified();
+
+  // In a stackified code register liveness has no meaning.
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  MRI.invalidateLiveness();
+}
+
+void EVMStackifyCodeEmitter::visitCall(const CFG::FunctionCall &Call) {
+  // Validate stack.
+  assert(Emitter.getStackHeight() == static_cast<int>(CurrentStack.size()));
+  assert(CurrentStack.size() >= Call.NumArguments + (Call.CanContinue ? 1 : 0));
+
+  // Assert that we got the correct return label on stack.
+  if (Call.CanContinue) {
+    [[maybe_unused]] const auto *returnLabelSlot =
+        std::get_if<FunctionCallReturnLabelSlot>(
+            &CurrentStack.at(CurrentStack.size() - Call.NumArguments - 1));
+    assert(returnLabelSlot && returnLabelSlot->Call == Call.Call);
+  }
+
+  // Emit code.
+  const MachineOperand *CalleeOp = Call.Call->explicit_uses().begin();
+  assert(CalleeOp->isGlobal());
+  Emitter.emitFuncCall(Call.Call, CalleeOp->getGlobal(),
+                       Call.Call->getNumExplicitDefs() - Call.NumArguments -
+                           (Call.CanContinue ? 1 : 0),
+                       Call.CanContinue);
+
+  // Update stack, remove arguments and return label from CurrentStack.
+  for (size_t I = 0; I < Call.NumArguments + (Call.CanContinue ? 1 : 0); ++I)
```

Comment:
`CurrentStack.erase`?

Reply:
Done. This is now part of the `adjustStackForInst` function.